src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,337 @@
+    ReturnErrorOnFailure(epProvider->ServerClusterInterfaces(serverClusterBuilder));
+    auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+    // Ensure capacity before appending.
+    ReturnErrorOnFailure(out.EnsureAppendCapacity(serverClusters.size()));
+
+    for (const auto * serverCluster : serverClusters)
+    {
+        if (!serverCluster->GetPaths().empty())
+        {
+            auto path = serverCluster->GetPaths().front();
+            ReturnErrorOnFailure(
+                out.Append({ path.mClusterId, serverCluster->GetDataVersion(path), serverCluster->GetClusterFlags(path) }));
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::GeneratedCommands(const ConcreteClusterPath & path, ReadOnlyBufferBuilder<CommandId> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->GeneratedCommands(path, out);
+}
+CHIP_ERROR CodeDrivenDataModelProvider::AcceptedCommands(const ConcreteClusterPath & path,
+                                                         ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->AcceptedCommands(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Attributes(const ConcreteClusterPath & path,
+                                                   ReadOnlyBufferBuilder<DataModel::AttributeEntry> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->Attributes(path, out);
+}
+
+void CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged(const AttributePathParams & path)
+{
+    ChipLogDetail(DataManagement, "CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged");
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::AddEndpoint(EndpointProviderRegistration & registration)
+{
+    ReturnErrorOnFailure(mEndpointProviderRegistry.Register(registration));
+
+    // If the provider has already been started, we need to startup the clusters for this new endpoint.
+    if (mServerClusterContext.has_value())
```

Comment:
invert the condition to allow for early return, so we decrease the indent. Something like

```cpp
VerifyOrReturnError(mServerClusterContext.has_value(), CHIP_NO_ERROR);
```